File: python/hashes/1711-count-good-meals.py
from typing import List


def countPairs(self, deliciousness: List[int]) -> int:
    counter = Counter(deliciousness)
    total = 0
    
    for d in counter:
        for i in range(22):
            t = (1 << i) - d
            if t in counter:
                if t == d:
                    total += (counter[t] - 1) * counter[t]
                else:
                    total += counter[t] * counter[d]
    
    return (total // 2) % (10**9 + 7)

File: cpp/hashes/1711-count-good-meals.cpp
#include <iostream>
#include <unordered_map>
#include <vector>

using namespace std;


int countPairs(vector<int>& deliciousness) {
    unordered_map<int, int> counter;
    int total = 0;

    for(int d: deliciousness){
        for(int i = 0; i < 22; i++){
            int p = 1 << i;
            int y = p - d;
            if(counter.count(y)) {
                total += counter[y];
                total %= 1e9 + 7;
            }
        }
        counter[d]++;
    }
    return total;
}